ImperatorToCK3/Source/CK3/CK3World.cpp

```diff
@@ -240,88 +240,80 @@ std::optional<std::pair<unsigned long long, std::shared_ptr<Imperator::Province>
 
 void CK3::World::linkCountiesToTitleHolders(const Imperator::World& impWorld)
 {
-	for (const auto& [name, landedTitle] : landedTitles.getFoundTitles())
+	for (const auto& [name, title] : getTitles())
 	{
-		if (name.find("c_")==0) // title is a county
+		if (name.find("c_")==0 && title->capitalBaronyProvince > 0) // title is a county and its capital province has a valid ID (0 is not a valid province in CK3)
 		{
-			auto countyTitle = std::make_shared<Title>();
-			countyTitle->titleName = name;
-			const auto capitalBaronyProvince = landedTitle.capitalBaronyProvince;
-
-			if (capitalBaronyProvince >0) // 0 is not a valid province in CK3
-			{
-				if (provinces.find(capitalBaronyProvince)==provinces.end())
-					LOG(LogLevel::Warning) << "Capital barony province not found " << capitalBaronyProvince;
+				if (provinces.find(title->capitalBaronyProvince)==provinces.end())
+					LOG(LogLevel::Warning) << "Capital barony province not found " << title->capitalBaronyProvince;
 				else
 				{
-					auto impProvince = provinces.find(capitalBaronyProvince)->second->srcProvince;
+					auto impProvince = provinces.find(title->capitalBaronyProvince)->second->srcProvince;
 					if (impProvince)
 					{
 						std::optional<unsigned long long> impMonarch;
 						if (impWorld.getCountries().find(impProvince->getOwner()) != impWorld.getCountries().end()) impMonarch = impWorld.getCountries().find(impProvince->getOwner())->second->getMonarch();
-						if (impMonarch) countyTitle->holder = "imperator" + std::to_string(*impMonarch);
+						if (impMonarch)
+						{
+							title->holder = "imperator" + std::to_string(*impMonarch);
+							countyHoldersCache.insert(title->holder);
+						}
 					}
 					else // county is probably outside of Imperator map
 					{
 						auto vanillaHistory = titlesHistory.popTitleHistory(name);
-						if (titlesHistory.currentHolderIdMap[name]) countyTitle->holder = *titlesHistory.currentHolderIdMap[name];
-						if (vanillaHistory) countyTitle->historyString = *vanillaHistory;
+						if (titlesHistory.currentHolderIdMap[name])
+						{
+							title->holder = *titlesHistory.currentHolderIdMap[name];
+							countyHoldersCache.insert(title->holder);
+						}
+
+						auto liegePtr = *titlesHistory.currentLiegeIdMap[name];
+						const auto dfLiegeName = titlesHistory.currentLiegeIdMap[name];
+						if (dfLiegeName && getTitles().find(*dfLiegeName) != getTitles().end())
+							title->setDeFactoLiege(getTitles().find(*dfLiegeName)->second);
+							
+						if (vanillaHistory) title->historyString = *vanillaHistory;
 					}
 				}
-			}
-			titles.insert(std::pair(name, countyTitle));
 		}
 	}
 }
 
 
-void CK3::World::importVanillaNonCountyNonBaronyTitles(const Imperator::World& impWorld)
+void CK3::World::vanillaNonCountyNonBaronyTitlesDetails(const Imperator::World& impWorld)
```

Comment:
Functions names should have both a noun and a verb. What are you doing with the vanillaNonCountyNonBaronyTitlesDetails?

Reply:
Actually merged it with `linkCountiesToTitleHolders` into `addHoldersAndHistoryToTitles`.